### PR TITLE
[RLP 1] Module testing and clean-up

### PR DESCRIPTION
**Pull request by AlexandraRoatis**
## Description

- added unit tests based on [rlptest.json](https://github.com/ethereum/tests/blob/develop/RLPTests/rlptest.json) and additional examples I considered relevant
- removed some unused & untested functionality from `RLP.java`
- optimized the `CompactEncoder.java` functionality to reduce the number of created byte arrays
- some other edge case unit tests

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] Bug fix.
- [ ] New feature.
- [x] Enhancement.
- [x] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

- existing test suite and added unit tests
- also cherry-picked these changes to the master branch and ran the kernel

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [x] Any dependent changes have been made.

### CODE REVIEW

File: modRlp/src/org/aion/rlp/CompactEncoder.java [hunk 1]
```diff
@@ -112,16 +110,35 @@
...
+                ret[0] = (byte) flag;
+                ret[1] = 0;
+                System.arraycopy(nibbles, 0, ret, 2, nibbles.length);
+            } else { // terminator == 1
+                ret = new byte[nibbles.length + 1];
+                ret[0] = (byte) flag;
+                ret[1] = 0;
+                System.arraycopy(nibbles, 0, ret, 2, nibbles.length - 1);
+            }
+            nibbles = ret;
```
Comment (AionJayT) [inline modRlp/src/org/aion/rlp/CompactEncoder.java]: This method seems like can be refactored cause the nibbles been copied twice during the logic (including the following ByteArrayOutputStream write). Considering find a way to avoid too many copies in this method.

File: modRlp/build.xml [hunk 2]
```diff
@@ -16,6 +16,9 @@
   <path id="classpath.test">
     <pathelement location="${dir.lib}/junit_4/junit-4.12.jar"/>
     <pathelement location="${dir.lib}/hamcrest/hamcrest-all-1.3.jar"/>
+    <pathelement location="${dir.lib}/truth-0.42.jar"/>
+    <pathelement location="${dir.lib}/guava-25.1-jre.jar"/>
+    <pathelement location="${dir.lib}/diffutils-1.3.0.jar"/>
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/build.xml]: I think we should. The versions we are using are quite outdated.

File: modRlp/src/org/aion/rlp/CompactEncoder.java (hunk 1 above)
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/CompactEncoder.java]: optimized in PR #572 commit 50ce43b

File: modRlp/src/org/aion/rlp/RLP.java [hunk 3]
```diff
@@ -647,23 +458,24 @@ public static DecodeResult decode(byte[] data, int pos) {
...
-                                                                        // its
-                                                                        // own
-                                                                        // RLP
-                                                                        // encoding
-        } else if (prefix < OFFSET_LONG_ITEM) {
+            return new DecodeResult(pos + 1, new byte[] {data[pos]}); // byte
+            // is
+            // its
+            // own
+            // RLP
```
Comment (aion-kelvin) [inline modRlp/src/org/aion/rlp/RLP.java]: awkward

File: modRlp/src/org/aion/rlp/Utils.java [hunk 4]
```diff
@@ -66,10 +79,14 @@
...
     static byte[] hexEncode(byte[] in) {
+        return hexEncode(in, false);
+    }
+
+    static byte[] hexEncode(byte[] in, boolean withTerminatorByte) {
         if (in == null) {
             return null;
         }
-        byte[] ret = new byte[in.length * 2];
+        byte[] ret = new byte[in.length * 2 + (withTerminatorByte ? 1 : 0)];
```
Comment (aion-kelvin) [inline modRlp/src/org/aion/rlp/Utils.java]: Why do we add that extra byte at the end that never gets populated?  (Don't think the loop below ever writes into ret[in.length*2+1])

File: modRlp/test/org/aion/rlp/UtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    public void testHexEncode_wSingleByte() {
+        for (byte b : Utils.encodingTable) {
+            byte[] input = new byte[] {b};
+            assertThat(Utils.hexEncode(input)).isEqualTo(Hex.encode(input));
+        }
+    }
+
+    @Test
+    public void testHexEncode_wString() {
+        String value = "1234567890abcdef";
```
Comment (aion-kelvin) [inline modRlp/test/org/aion/rlp/UtilsTest.java]: Not directly related to your code, but I was experimenting this with test and I added an 'a' to this string, so value = "1234567890abcdefa" and the Hex.decode(value) throws an exception.  Is that expected?

File: modRlp/test/org/aion/rlp/RLPTest.java [hunk 6]
```diff
@@ -95,41 +89,18 @@
...
-        byte[] payload = Hex.decode(peersPacket);
-
-        byte[] ip = decodeIP4Bytes(payload, 5);
-
-        assertEquals(InetAddress.getByAddress(ip).toString(), ("/54.204.10.41"));
-    }
-
-    @Test
-    public void test2() throws UnknownHostException {
+    public void test2() {
```
Comment (aion-kelvin) [inline modRlp/test/org/aion/rlp/RLPTest.java]: Optional: maybe rename this test() or test1()

File: modRlp/test/org/aion/rlp/RLPSpecExtraTest.java [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+import org.aion.base.util.Hex;
+import org.junit.Test;
+
+/**
+ * Tests for the RLP implementation that are complementary to RLP specification tests from Ethereum
+ * implemented in {@link RLPSpecTest}.
+ *
+ * @author Alexandra Roatis
+ */
+public class RLPSpecExtraTest {
```
Comment (aion-kelvin) [inline modRlp/test/org/aion/rlp/RLPSpecExtraTest.java]: Looks good thanks for the extra tests, this looks pretty well covered.

One question -- I didn't see any numerical tests where the input is a negative number.  Is that supposed to be possible?  To encode a negative number, then decode the result, and get back the same negative number?

File: modRlp/src/org/aion/rlp/CompactEncoder.java [hunk 8]
```diff
@@ -31,56 +31,54 @@
...
  *
- * An alternative way of thinking about this to not think of there being a
- * terminator symbol, but instead treat bit specifying the existence of the
- * terminator symbol as a bit specifying that the given node encodes a final
- * node, where the value is an actual value, rather than the hash of yet another
- * node.
+ * <p>To solve both of these issues, we force the first nibble of the final byte-stream to encode
+ * two flags, specifying oddness of length (ignoring the 'T' symbol) and terminator status; these
+ * are placed, respectively, into the two lowest significant bits of the first nibble. In the case
+ * of an even-length hex string, we must introduce a second nibble (of value zero) to ensure the
```
Comment (arajasek) [inline modRlp/src/org/aion/rlp/CompactEncoder.java]: I understand what it's saying, but I find this last sentence quite confusingly worded. If it's not a bother, I'd suggest rephrasing it so that it doesn't sound like "we must ensure that even-length hex strings are even-length hex strings".

File: modRlp/src/org/aion/rlp/CompactEncoder.java (hunk 1 above)
Comment (AionJayT) [reply modRlp/src/org/aion/rlp/CompactEncoder.java]: good job!

File: modRlp/src/org/aion/rlp/RLP.java [hunk 9]
```diff
@@ -396,47 +193,60 @@ public static void fullTraverse(byte[] msgData, int level, int startPos, int end
                     int length = calcLength(lengthOfLength, msgData, pos);
 
                     // now we can parse an item for data[1]..data[length]
-                    System.out.println("-- level: [" + level + "] Found big list length: " + length);
+                    System.out.println(
+                            "-- level: [" + level + "] Found big list length: " + length);
```
Comment (AionJayT) [inline modRlp/src/org/aion/rlp/RLP.java]: Does this be printed to the terminal during kernel running? If yes, can we add new logEnum for the RLP module?

File: modRlp/test/org/aion/rlp/RLPSpecExtraTest.java [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+
+    // ENCODING
+
+    @Test
+    public void testEncodeShort3() {
```
Comment (aionick) [inline modRlp/test/org/aion/rlp/RLPSpecExtraTest.java]: Can we also test the max value for `short` and the min values for `short`,`int`,`long` (looks like our min is zero?) just to have those edges covered

File: modRlp/test/org/aion/rlp/UtilsTest.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+ */
+package org.aion.rlp;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import java.util.Arrays;
+import org.aion.base.util.Hex;
+import org.junit.Test;
+
+public class UtilsTest {
```
Comment (aionick) [inline modRlp/test/org/aion/rlp/UtilsTest.java]: Could we also have tests for concatenating two empty arrays as well as empty with non-empty arrays?

File: modRlp/src/org/aion/rlp/RLP.java (hunk 9 above)
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/RLP.java]: I removed this functionality in #572 because it wasn't used.

File: modRlp/src/org/aion/rlp/Utils.java (hunk 4 above)
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/Utils.java]: I did this as an optimization to avoid creating and copying several byte arrays. This way the proper length array gets created from the start. Since this is just a utility, I prefer keeping it as is.

File: modRlp/test/org/aion/rlp/DecodeResultTest.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+ *     If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Contributors:
+ *     Aion foundation.
+ */
+package org.aion.rlp;
+
+import org.junit.Test;
+
+public class DecodeResultTest {
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/DecodeResultTest.java]: The class is actually very well covered because of the `RLP.decode` calls.

File: modRlp/test/org/aion/rlp/RLPElementTest.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+import static com.google.common.truth.Truth.assertThat;
+
+import org.aion.base.util.ByteUtil;
+import org.junit.Test;
+
+public class RLPElementTest {
+
+    @Test(expected = RuntimeException.class)
+    public void testRLPList_wRuntimeException() {
+        RLPList.recursivePrint(null);
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/RLPElementTest.java]: The class is very well covered because of the `RLP.decode2` calls. Only the null checks were missing, and it felt more elegant to put both tests in one class instead of having a class for each with a single test. If you insist I can split them into individual classes. Do you? :)

File: modRlp/test/org/aion/rlp/RLPElementTest.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    }
+
+    @Test
+    public void testRLPItem_wNull() {
+        RLPItem item = new RLPItem(null);
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/RLPElementTest.java]: see comment above

File: modRlp/test/org/aion/rlp/RLPSpecExtraTest.java (hunk 7 above)
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/RLPSpecExtraTest.java]: Nope, the encoding does not work with negative numbers. They get interpreted as positive, so you would get a different value when decoding.

File: modRlp/test/org/aion/rlp/RLPSpecExtraTest.java (hunk 10 above)
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/RLPSpecExtraTest.java]: That's covered in the `RLPSpecTest.java` which is based on [rlptest.json](https://github.com/ethereum/tests/blob/develop/RLPTests/rlptest.json). These tests complement those test cases with values that I considered may be relevant.

File: modRlp/test/org/aion/rlp/UtilsTest.java (hunk 5 above)
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/UtilsTest.java]: `Hex.decode(value)` works only with even length strings because one byte is encoded to two characters.

File: modRlp/test/org/aion/rlp/RLPElementTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (aion-kelvin) [reply modRlp/test/org/aion/rlp/RLPElementTest.java]: Oh, I see, makes sense... we can leave it this way for now then.

Philosophically, for a class "X" like having a corresponding "XTest.java," but in this case I think it really doesn't make much difference.

File: modRlp/test/org/aion/rlp/RLPElementTest.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (aion-kelvin) [reply modRlp/test/org/aion/rlp/RLPElementTest.java]: Sounds good

File: modRlp/src/org/aion/rlp/CompactEncoder.java [hunk 15]
```diff
@@ -112,16 +110,35 @@
 
         if (nibbles[nibbles.length - 1] == TERMINATOR) {
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/CompactEncoder.java]: done in PR #572 commit ba5c2954639c95eee8258f5d9c4392d1ad147225

File: modRlp/test/org/aion/rlp/CompactEncoderTest.java [hunk 16]
```diff
@@ -31,93 +31,115 @@
...
-import org.junit.Test;
-
+import static com.google.common.truth.Truth.assertThat;
 import static org.junit.Assert.assertArrayEquals;
 
-import org.aion.rlp.CompactEncoder;
+import java.util.Arrays;
+import org.junit.Test;
 
 public class CompactEncoderTest {
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/CompactEncoderTest.java]: done in PR #572 commit ba5c2954639c95eee8258f5d9c4392d1ad147225

File: modRlp/src/org/aion/rlp/RLP.java (hunk 3 above)
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/RLP.java]: fixed in PR #572 commit ba5c2954639c95eee8258f5d9c4392d1ad147225

File: modRlp/test/org/aion/rlp/RLPTest.java (hunk 6 above)
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/RLPTest.java]: gave it a meaningful name in PR #572 commit ba5c2954639c95eee8258f5d9c4392d1ad147225

File: modRlp/test/org/aion/rlp/UtilsTest.java (hunk 11 above)
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/UtilsTest.java]: done in PR #572 commit ba5c2954639c95eee8258f5d9c4392d1ad147225

File: modRlp/src/org/aion/rlp/CompactEncoder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/CompactEncoder.java]: improved in PR #572 commit ba5c2954639c95eee8258f5d9c4392d1ad147225
